apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts

```diff
@@ -48,19 +49,63 @@ ${context}
  *
  * @param client The OpenAI client to use.
  * @param prompt The user's question generation prompt.
+ * @param promptUserInput The user's indication of how to create student input boxes.
+ * @param mandatoryElementNames Elements that we must pull documentation for.
  * @param authnUserId The user's authenticated user ID.
  * @returns A string of all relevant context documents.
  */
-async function makeContext(client: OpenAI, prompt: string, authnUserId: string): Promise<string> {
+async function makeContext(
+  client: OpenAI,
+  prompt: string,
+  promptUserInput: string | undefined,
+  mandatoryElementNames: string[],
+  authnUserId: string,
+): Promise<string> {
   const embedding = await createEmbedding(client, prompt, openAiUserFromAuthn(authnUserId));
 
+  //identify all elements that we are using *and* have documentation document chunks.
+  const mandatoryElements =
+    mandatoryElementNames.length > 0
+      ? await queryRows(
+          sql.select_documents_by_chunk_id,
+          {
+            doc_path: 'docs/elements.md',
+            chunk_ids: mandatoryElementNames,
+          },
+          QuestionGenerationContextEmbeddingSchema,
+        )
+      : [];
+
+  const numElements = Math.max(5 - mandatoryElements.length, 0);
+
   const docs = await queryRows(
     sql.select_nearby_documents,
-    { embedding: vectorToString(embedding), limit: 5 },
+    { embedding: vectorToString(embedding), limit: numElements },
     QuestionGenerationContextEmbeddingSchema,
   );
 
-  return docs.map((doc) => doc.doc_text).join('\n\n');
+  //if a prompt specifies how user input is handled, try to find documentation for those types of input
+  //and save as last doc. Regeneration prompts don't use this, so promptUserInput may be undefined.
+  if (promptUserInput !== undefined) {
```

Comment:
I don't 100% understand why this is here. Is this to ensure that we have documentation about a least one possibly-relevant elements? Could we generalize this to something like  this:

```ts
if (mandatoryElements.length === 0 && !docs.some(d => d.doc_path === 'docs/elements.md')) {
  // We don't have any element documentation, add some
}
```

Reply:
the idea is that the embeddings of the user's description on *how* users should enter their input should be generated should be closer to relevant docs than the embeddings of the entire prompt. I do agree that we want to generalize further, but just providing for always one element being documentation doesn't capture what I was going for.

Reply:
Doesn't this still use the embedding of the prompt as a whole though?

Reply:
it shouldn't have; changed in a95d2e5c4e4d6c01f8b0919921ab09d2dbe5caba. i should have read my old code before responding 🤦‍♂️